src/nupic/algorithms/SDRClassifier.cpp

```diff
@@ -0,0 +1,745 @@
+          if (!actualValuesSet_[bucketIdx] || category)
+          {
+            actualValues_[bucketIdx] = actValue;
+            actualValuesSet_[bucketIdx] = true;
+          } else {
+            actualValues_[bucketIdx] =
+                ((1.0 - actValueAlpha_) * actualValues_[bucketIdx]) +
+                (actValueAlpha_ * actValue);
+          }
+
+          // compute errors and update weights
+          deque<vector<UInt>>::const_iterator patternIteration =
+              patternNZHistory_.begin();
+          for (deque<UInt>::const_iterator learnIteration =
+               iterationNumHistory_.begin();
+               learnIteration !=iterationNumHistory_.end();
+               learnIteration++, patternIteration++)
+          {
+            const vector<UInt> learnPatternNZ = *patternIteration;
+            UInt nSteps = learnIteration_ - *learnIteration;
+
+            // update weights
+            if (binary_search(steps_.begin(), steps_.end(), nSteps))
+            {
+              vector<Real64> error = calculateError_(bucketIdx, 
+                patternNZ, nSteps);
+              for (auto& bit : learnPatternNZ)
+              {
+                for (UInt bucket = 0; bucket < maxBucketIdx_; ++bucket)
+                {
+                  weightMatrix_[nSteps].at(bit, 
+                    bucket) += alpha_ * error[bucket];
+                }
+              }
+            }
+          }
+
+        }
+
+      }
+
+      UInt SDRClassifier::persistentSize() const
+      {
+        stringstream s;
+        s.flags(ios::scientific);
+        s.precision(numeric_limits<double>::digits10 + 1);
+        save(s);
+        return s.str().size();
+      }
+
+      void SDRClassifier::infer_(const vector<UInt>& patternNZ, UInt bucketIdx, 
+        Real64 actValue, ClassifierResult* result)
+      {
+        // add the actual values to the return value. For buckets that haven't
+        // been seen yet, the actual value doesn't matter since it will have
+        // zero likelihood.
+        vector<Real64>* actValueVector = result->createVector(
+            -1, actualValues_.size(), 0.0);
+        for (UInt i = 0; i < actualValues_.size(); ++i)
+        {
+          if (actualValuesSet_[i])
+          {
+            (*actValueVector)[i] = actualValues_[i];
+          } else {
+            // if doing 0-step ahead prediction, we shouldn't use any
+            // knowledge of the classification input during inference
+            if (steps_.at(0) == 0)
+            {
+              (*actValueVector)[i] = 0;
+            } else {
+              (*actValueVector)[i] = actValue;
+            }
+          }
+        }
+
+        for (auto nSteps = steps_.begin(); nSteps!=steps_.end(); ++nSteps)
+        {
+          vector<Real64>* likelihoods = result->createVector(*nSteps, 
+            maxBucketIdx_ + 1, 1.0 / actualValues_.size());
+          for (auto& bit : patternNZ)
+          {
+            Matrix weights = weightMatrix_[*nSteps];
```

Comment:
Also, use `.at` to avoid the default behavior of creating a new Matrix when the key isn't found. I think this will work:

`const Matrix& weights = weightMatrix_.at(*nSteps);`